sale_loyalty_order_line_link/models/sale_order.py

```diff
@@ -176,9 +151,9 @@ def _filter_related_program_lines(self, program):
         the product criteria rules, we can extend this method to return the proper
         records."""
         # No domain means that any product meets the promotion rules
-        if not program.rule_products_domain:
-            return self
-        domain = safe_eval(program.rule_products_domain)
+        if not program.program_id.rule_ids.product_domain:
+            return self.filtered(lambda x: not x.is_reward_line)
+        domain = safe_eval(program.program_id.rule_ids.product_domain)
```

Comment:
This is going to crash for sure if there are several rules. But you've got a more reliable way to obtain the domain so you can combine them: https://github.com/odoo/odoo/blob/adae60fc532cf6a853815f840cb49e93eb008cb7/addons/loyalty/models/loyalty_rule.py#L117-L128

Reply:
I've gone through the whole code, correcting nomenclatures and o2m field accesses to avoid errors and one of the changes is applied in this part.